### PR TITLE
Django IDOM

### PR DESCRIPTION
**Pull request by Archmonger**
This is the most minimalistic Django configuration possible that can support Websockets.

I've used folder structure and file naming schemes suggested by Django documentation.

Also included the Daphne webserver in the requirements for testing on more than just the development webserver.

In order to get this configuration running
1) `cd <repo_root_dir>`
2) `pip install requirements.txt`
3) `python manage.py migrate` to create the initial database
4) `python manage.py runserver` to run the django development test server. Alternatively, `daphne dj_idom.asgi:application` to run the  production-grade webserver.

Here's the files in the repo:
`dj_idom/static/scripts.js`: Client side websocket
`dj_idom/templates/base.html`: HTML base template.
`dj_idom/consumers.py`: Server side websocket
`dj_idom/asgi.py`: Websocket URL routing
`dj_idom/urls.py`: HTTP URL routing
`dj_idom/settings.py`: Django boot time config
`manage.py`: Django project management utility

### CODE REVIEW

**Comment by rmorshea** [comment]
Thanks, I'll look this over shortly.

**Comment by rmorshea** [comment]
@Archmonger IDOM 0.29.0 was just released and contains a fix so that we won't need to do `pip install idom[stable]` here.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+          node-version: "14"
+      - name: Use Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install Python Dependencies
+        run: pip install -r requirements/test-run.txt
+      - name: Run Tests
+        run: |
+          npm install -g npm@v7.13.0
```
Comment (rmorshea) [inline .github/workflows/test.yml:33]: I'm not sure why I pinned the version of NPM here. I'd rather get an error when something no longer works with the latest version of NPM. So if we could just swap this line out for the following that would be great:

```
npm install -g npm@latest
npm --version
```

**Comment by Archmonger** [comment]
@rmorshea Resolved all startup errors under this repo's default configuration, I'll look to implement some tests soon.

**Comment by Archmonger** [comment]
@rmorshea Getting a test_suite fail on GH actions, not sure what to do to resolve it.

**Comment by rmorshea** [comment]
Yup, I think a test that goes something like this should be totally sufficient:

```python
import idom

def test_simple_counter():
    run_the_app_somehow()
    web_driver = make_the_web_driver_somehow()


    @idom.component
    def Counter():
        count, set_count = idom.hooks.use_state(0)
        return idom.html.div(
            idom.html.button(
                {"id": "incr", "onClick": lambda event: set_count(count + 1)},
                "Click me!",
            ),
            idom.html.p(
                {"id": "count", "data-count": count}, f"Current count is: {count}"
            )
        )
    
    counter_button = web_driver.find_element_by_id("incr")
    counter_view = web_dirver.find_element_by_id("count")

    for i in range(5):
        assert counter_view.get_attribute("data-count") == i
        counter_button.click()
```

Anything beyond that is nice but probably not necessary right now.

**Comment by Archmonger** [comment]
I think I'll need to tag up with you regarding the frontend-backend connection via django. Not sure how to get those components to actually appear in a django fashion.

**Comment by rmorshea** [comment]
I think there's some docs on on how to set up a test using Selenium: https://docs.djangoproject.com/en/3.2/topics/testing/tools/#liveservertestcase

If there's not enough info there, grab a weekly office hour slot and we'll figure it out: https://github.com/idom-team/idom/discussions/312

**Comment by rmorshea** [comment]
There's some critical bugs in IDOM that need attention. Probably won't have time to work on this for another week or so.

**Comment by Archmonger** [comment]
Does the [commit from 6 days ago](https://github.com/idom-team/idom/commit/b7e729247b622a390f08fb8f1cc1bf23aa9ecad6) address the critical bugs you brought up?

**Comment by rmorshea** [comment]
It does. I plan to release it soon. I'll then have some time to return to this after the two remaining open PRs are merged..

**Comment by rmorshea** [comment]
Ok, I'm not free to start working on this again.

**Comment by Archmonger** [comment]
> Ok, I'm not free to start working on this again.

😂

I've been focused on Pyarr development but I'll have time to return to this soonish

**Comment by rmorshea** [comment]
Edit: I'm **now** free to start working on this again.

**Comment by Archmonger** [comment]
Can't wait to test this out!! This is the last piece of the puzzle for developing the Conreq community app store!

**Review by rmorshea — APPROVED** [review]

**Comment by rmorshea** [comment]
I'm gonna go ahead and merge this, but I really do think we should provide an app that encapsulates what was done in `test_app` since there's a lot of little things you have to understand about IDOM to make it work. Unfortunately, I don't think I understand enough about Django to make that happen.

Ideally one would be able to install `django_idom` as an app and then just embed a template into you HTML where you want an IDOM view to show up. I think I'll have to do some pair programming with someone who knows how to make re-usable apps though.

**Comment by Archmonger** [comment]
> I'm gonna go ahead and merge this, but I really do think we should provide an app that encapsulates what was done in `test_app` since there's a lot of little things you have to understand about IDOM to make it work. Unfortunately, I don't think I understand enough about Django to make that happen.
> 
> Ideally one would be able to install `django_idom` as an app and then just embed a template into you HTML where you want an IDOM view to show up. I think I'll have to do some pair programming with someone who knows how to make re-usable apps though.

I'll can show you how to make reusable templates. That one is fairly simple, but it might be awkward given the current repo structure of using test_app

**Comment by rmorshea** [comment]
@Archmonger it'd be great to organize another working session to get through that.

**Comment by Archmonger** [comment]
> @Archmonger it'd be great to organize another working session to get through that.

We can plan something out around your schedule, let me know your availability.

**Comment by Archmonger** [comment]
@rmorshea Something came up so I'll need to cancel our 11. I'll be able to do later in the evening if you're free.

**Comment by rmorshea** [comment]
Should be fine. Just let me know when you're able.

**Comment by rmorshea** [comment]
@Archmonger still able to meet?

**Comment by Archmonger** [comment]
@rmorshea just freed up, lemme know when you wanna meet

**Comment by rmorshea** [comment]
Moved the event.

**Pull request merged**